### PR TITLE
Support arbitrary stream merges

### PR DESCRIPTION
**Pull request by fstirlitz**
This supersedes #5581 and #6435.

With this change, the merge operator may join any number of media streams, video or audio. The streams are downloaded in the order specified.

Use case:

```
youtube-dl http://www.gdcvault.com/play/1014631/Classic-Game-Postmortem-PAC -f speaker+slides+en
```

shall download the `slides` (video only), `speaker` (video+audio) and `en` (audio only) streams and merge them into a single file.

Remarks:
- If more than one video or audio streams are to be merged, the default output format is set to `mkv`. This isn't always necessary: `webm`, `ogg` and `mp4` will do fine with multiple audio and/or video streams; `flv`, on the other hand, will not (and possibly others).
- The metadata for a stream of a particular type are copied into the `info_dict` of the merged stream only if there is a single stream of that type (i.e. only one audio stream and only one video stream). This seems sensible (after all, questions like "what is the audio codec of `bestaudio[acodec=opus]+bestaudio[acodec=vorbis]`" are somewhat ill-formed), but other behaviours might be preferable here (e.g. copying the metadata of the "best-quality" stream of a given type, or copying them if they are the same for every stream and leaving them otherwise, or creating a list of all the values for all the streams).

### CODE REVIEW

File: youtube_dl/YoutubeDL.py [hunk 1]
```diff
@@ -1511,17 +1531,21 @@ def dl(name, info):
...
-                            COMPATIBLE_EXTS = (
-                                ('mp3', 'mp4', 'm4a', 'm4p', 'm4b', 'm4r', 'm4v'),
-                                ('webm')
-                            )
-                            for exts in COMPATIBLE_EXTS:
-                                if video_ext in exts and audio_ext in exts:
-                                    return True
+                        exts = set(format.get('ext') for format in formats)
+                        COMPATIBLE_EXTS = (
+                            set('mp3', 'mp4', 'm4a', 'm4p', 'm4b', 'm4r', 'm4v'),
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]:
```
Traceback (most recent call last):
  File "youtube_dl/__main__.py", line 19, in <module>
    youtube_dl.main()
  File "C:\Dev\git\youtube-dl\master\youtube_dl\__init__.py", line 410, in main
    _real_main(argv)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\__init__.py", line 400, in _real_main
    retcode = ydl.download(all_urls)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1649, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 669, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 715, in process_ie_result
    return self.process_video_result(ie_result, download=download)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1315, in process_video_result
    self.process_info(new_info)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1558, in process_info
    if self.params.get('merge_output_format') is None and not compatible_formats(requested_formats):
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1543, in compatible_formats
    set('mp3', 'mp4', 'm4a', 'm4p', 'm4b', 'm4r', 'm4v'),
TypeError: set expected at most 1 arguments, got 7
```
What's the point at all?

**Comment by fstirlitz** [comment]
Anyone willing to merge this?

**Comment by yan12125** [comment]
From #5154: cases like audio+audio or video+video should be implemented as well

**Comment by fstirlitz** [comment]
**THEY ARE.** What part of 'arbitrary' is unclear to you? The only caveat is that metadata is not copied to the parent `info_dict` in such cases.

**Comment by fstirlitz** [comment]
Is there a reason why this isn't merged yet?

**Pull request closed without merging**